### PR TITLE
PR Handling Frontend Login and Register

### PR DESCRIPTION
**Pull request by AnkushSharma2698**
In this PR:

1. Set up Redux state across the application
2. Error Handling and Auth state
3. Connection to backend for storing tokens and handling login from frontend to backend to retrieve token.
4. Consistent Checking per request to ensure authenticated.
5. Since every route is protected in the BE, the way I handle auth may change but it seems to be fine for now

#42

### CODE REVIEW

File: frontend/src/actions/authActions.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import {
+    USER_LOADED,
+    USER_LOADING,
+    AUTH_ERROR,
+    LOGIN_SUCCESS,
+    LOGIN_FAIL,
+    LOGOUT_SUCCESS,
+    REGISTER_SUCCESS,
+    REGISTER_FAIL,
+} from './types';
```
Comment (PENGYUXIONG) [inline frontend/src/actions/authActions.js:14]: we had action types to store those, if you feels more comfortable having action types inside action folder then remove the old `actionType` folder

File: frontend/src/components/auth/RegisterForm.jsx [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        }
+        
+        // This should be authenticated routing
+        if (isAuthenticated) {
+            // Send the user to the homepage
+            console.log('Sending user to the homepage');
+        }
+    }
+
+    componentWillUnmount() {
```
Comment (PENGYUXIONG) [inline frontend/src/components/auth/RegisterForm.jsx:41]: I am not sure if react lifecycle will follow the intended order. 

Please make sure the props are working as you expected. Using lifecycle is a good way to update dynamic data but must be careful

File: frontend/src/reducers/authReducer.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+};
+
+
+export default function(state=initialState, action) {
+    switch(action.type) {
+        case USER_LOADING:
+            // Get user from backend
+            return {
+                ...state,
+                isLoading: true
```
Comment (PENGYUXIONG) [inline frontend/src/reducers/authReducer.js:27]: IS there going to have a loading page? Please explain a bit why we need this prop in state

File: frontend/src/components/auth/Logout.jsx [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import {connect} from 'react-redux';
+import {logout} from '../../actions/authActions';
+import { Menu } from 'antd';
+import { Redirect} from 'react-router-dom';
+class Logout extends React.Component {
+    render() {
+        
+        let {logout: logouter, ...rest} = this.props;
```
Comment (PENGYUXIONG) [inline frontend/src/components/auth/Logout.jsx:9]: Is this a good way to update the attributes in state?

I thought there should have some other ways to check.

File: frontend/src/actions/authActions.js (hunk 1 above)
Comment (AnkushSharma2698) [reply frontend/src/actions/authActions.js:14]: yeah I felt like one file is good, all of our actions live in one file

File: frontend/src/components/auth/RegisterForm.jsx (hunk 2 above)
Comment (AnkushSharma2698) [reply frontend/src/components/auth/RegisterForm.jsx:41]: yep this works because when we switch components it causes a re-render so the errors are cleared

File: frontend/src/reducers/authReducer.js (hunk 3 above)
Comment (AnkushSharma2698) [reply frontend/src/reducers/authReducer.js:27]: it allows for us to support loading if needed. I added it if we want to have loading spinners and stuff for styling. It is not currently used but is there for extensibility

File: frontend/src/components/auth/Logout.jsx (hunk 4 above)
Comment (AnkushSharma2698) [reply frontend/src/components/auth/Logout.jsx:9]: yeah, because the state is app wide, and all relevant components are checking on render for authentication, as soon as logout button is clicked and state is cleared, no matter what component we are at, we will be redirected to the landing page

Comment (PENGYUXIONG) [reply frontend/src/components/auth/Logout.jsx:9]: Sorry, Let me rephrase. There should have other ways to update the attributes in state. Directly pushing props into the state in component is an approach I have never seen before. 

If you feel good about this then ignore this msg. But I would suggest to have functions inside components for state updates.

**Review by PENGYUXIONG — APPROVED** [review]
Looks good to me! Just some structure advices

If you have tested it and worked, then it is all good.
